src/mongocrypt.c

```diff
@@ -440,6 +632,16 @@ mongocrypt_destroy (mongocrypt_t *crypt)
    bson_free (crypt->crypto);
    _mongocrypt_cache_oauth_destroy (crypt->cache_oauth_azure);
    _mongocrypt_cache_oauth_destroy (crypt->cache_oauth_gcp);
+
+#ifndef __linux__
+   mcr_dll_close (crypt->csfle_lib);
+#else
+   /// NOTE: On Linux, skip closing the CSFLE library itself, since a bug in the
+   /// way ld-linux and GCC interact causes static destructors to not run during
+   /// dlclose(). Still, free the error string that may be non-null:
+   mstr_free (crypt->csfle_lib.error_string);
```

Comment:
This is making me a bit nervous. :) Do you know anything more about what happens here? Generally speaking, static destructors *do* run for shared libraries compiled with GCC just fine on Linux. Without knowing more details, my gut feeling is that the server team compiles the shared library in a special way that keeps the shared library destructors from running, but in that case I think that should be reported to them as a bug.

Reply:
I consulted around with other devs and with the Internet, and we cannot figure out what is going on that causes the resource leaks. If the library is never unloaded (i.e. never call `dlcose()`), the static destructors _will_ run at program shutdown, and `dlopen()` is smart enough to never load a duplicate of the library, so I figure that this is the safest way to get destructors to definitely run without leaking, even though it is suboptimal.

Reply:
@vector-of-bool Yeah, this isn’t a blocking concern for sure. It’s still something that appears to be specific to the this specific shared library, iiuc, so maybe we could open a SERVER ticket? (I can also do that but it sounds like you already did a lot more digging.)

Reply:
I created a ticket to track this: [SERVER-63710](https://jira.mongodb.org/browse/SERVER-63710)